Update client-participation dependencies, and get off gulp -> webpack

metasoarous opened the pull request:
Pulling in @brendanarnold's work on https://github.com/sirodoht/suburb/pull/3 (see #1575) to update participation client.

ballPointPenguin committed:
some file deletions

ballPointPenguin committed:
more removal

ballPointPenguin committed:
changes from Chris's branch

ballPointPenguin committed:
revert some removals

ballPointPenguin committed:
many changes from suburb

ballPointPenguin committed:
troubleshooting webpack build

ballPointPenguin committed:
try upstream backbone

metasoarous committed:
get file-server/Dockerfile client-participation node/npm versions up to date

metasoarous committed:
Add missing require for CurrentUserModel

metasoarous committed:
Add missing jquery require

metasoarous committed:
Add back Google Analytics

ballPointPenguin committed:
update client p Dockerfile

ballPointPenguin committed:
restore some rtl js lines

ballPointPenguin committed:
remove unused var

ballPointPenguin committed:
remove old index.html

ballPointPenguin committed:
update client p README

ballPointPenguin commented:
A few things to keep in mind while troubleshooting or reviewing these changes:

- Backbone is now pointing to the [original upstream repo](https://github.com/jashkenas/backbone) (no more fork!)
- There are two versions of d3.js in client-participation: v3 (installed by npm) and v4 (minified local copy). Soon we should unify these. It seems v3 is only used by VisView.js
- Node and npm versions have been updated (and can probably be updated again to the latest stable)

metasoarous commented on client-participation/.dockerignore:
```diff
@@ -1,20 +1,8 @@
-.DS_Store
+.sass-cache/*
 *.log
-js/tmpl/
```
this is really weird... why was this ever in here!?

metasoarous replied:
not even sure how anything ever worked with this in here, since these files would be needed for compilation, right?

colinmegill replied:
Intermediate step in the build, only exists locally?

brendanarnold replied:
Yes, it used to be a two step process managed through Gulp, `.handlebars` templates would be compiled by Gulp into `.js` files in this directory and this is what the rest of the code imports - this probably ensures that the compiled js starts with a clean build directory and not just a copies of whatever is local. Now it is all handled through Webpack Handlebars loader in a single step.

metasoarous commented on client-participation/.dockerignore:
```diff
@@ -1,20 +1,8 @@
-node_modules
-bower_components
-public
-.DS_Store
```
Unsure on this; I think it's just a mac os filesystem thing, and I don't know how smart docker is about ignoring this by default. I'm less convinced it should be in the `.gitignore`, since mac devs should be responsible for having OS specific stuff like this in their global `~/.gitignore`. Then again, there's an argument that the project is more newby friendly if their are fewer of these gotchas, but I also don't want to overoptimize for that.

brendanarnold replied:
Its pretty common to see `.DS_Store` in `.gitignore` files just like its pretty common to see IDE specific ignores e.g. `.vscode`. There is next-to-no harm there being these extra entries and it does save some accidental commits and accidently importing some bogus cache files from other people's computers etc. It's in the `polis-whitelabel` fork for instance.

colinmegill replied:
Also prefer to keep these! Happy to have them wherever (global gitignore perhaps not participationView for instance) but we should be ignoring these.

colinmegill replied:
bower_components is just gone, hopefully, speaking here about things like `.DS_Store`

metasoarous commented:
Status update: Right now, the project builds and loads, but the vote buttons do not work. Nothing logs to either the browser console or the server, so it seems like the events just aren't firing.

This is the code in `client-participation/js/views/vote-view.js` that is supposed to be assigning event handler functions.

```
  events: {
    "click #agreeButton": "participantAgreed",
    "click #disagreeButton": "participantDisagreed",
    "click #passButton": "participantPassed",
    "click #subscribeBtn": "subscribeBtn",
    'submit #subscribeEmailForm': "subscribeBtn",
```

Unfortunately, there's not really a lot to go on here. However, the comment submission _does_ seem to work, so looking at that might be useful.

colinmegill commented on client-participation/webpack.config.js:
```diff
@@ -1,27 +1,271 @@
+                  // Expects lodash, Backbone and handlebars to be present
+                  'default handlebars Handlebars',
+                  'default backbone Backbone',
+                  'default lodash _'
```
I'm taking a look at this. This is a very old module and I'm not sure it can use lodash.

colinmegill replied:
https://github.com/pol-is/handlebones/blob/master/handlebones.js

colinmegill replied:
Looking at: https://medium.com/paypal-tech/from-require-js-to-webpack-part-2-the-how-1800917fa416

colinmegill replied:
https://github.com/walmartlabs/thorax

metasoarous replied:
@colinmegill Do you remember why we ended up having to maintain a fork of handlebones and handlebars?

colinmegill replied:
Here's the diff. The polis fork is on the right. 

<img width="1202" alt="image" src="https://user-images.githubusercontent.com/1770265/213101854-81c7112d-78fb-420d-9a8b-b83bec9d4beb.png">

...

<img width="1249" alt="image" src="https://user-images.githubusercontent.com/1770265/213101890-e3e9a9a2-f2df-4443-acfe-db8f98919199.png">

...

<img width="1314" alt="image" src="https://user-images.githubusercontent.com/1770265/213101998-45729edf-77e0-464e-9f05-9850dc1a4fee.png">

I read this change as safety checks after an encountered undefined case.

colinmegill replied:
See, especially: 

`if (nextEl)` 

which prevents `cannot read property nextSibling of undefined`

metasoarous commented:
Pinging @brendanarnold - were you able to get voting working, or have any suggestions as to what we might want to try here?

metasoarous committed:
point client-participation backbone dep to updated polis fork

merged in changes from sirodoht's fork: https://github.com/pol-is/backbone/pull/1

metasoarous committed:
switch to npm install in file-server client build process
